fdbrpc/include/fdbrpc/LoadBalance.actor.h

```diff
@@ -382,11 +389,17 @@ Future<Void> replicaComparison(Req req,
 			if (numMismatch) {
 				throw storage_replica_comparison_error();
 			} else if (((numError || numFetchReplicaTimeout) && (requiredReplicas == ALL_REPLICAS)) ||
-			           (successfulReplies != requiredReplicas && requiredReplicas > 0)) {
+			           (successfulReplies != requiredReplicas && requiredReplicas > 0 &&
+			            restOfTeamFutures.size() >= requiredReplicas)) {
 				const char* type = numError ? "ReplicaComparisonReadError" : "ReplicaComparisonTimeoutError";
-				TraceEvent(SevError, type).detail("SSError", replicaErrorCode);
-
-				throw Error(replicaErrorCode);
+				TraceEvent(SevWarnAlways, type)
+				    .detail("TeamSize", restOfTeamFutures.size() + 1)
+				    .detail("RequiredReplies", requiredReplicas)
+				    .detail("SuccessfulReplies", successfulReplies)
+				    .detail("SSError", replicaErrorCode);
+
+				throw Error((requiredReplicas == ALL_REPLICAS) ? replicaErrorCode
+				                                               : error_code_unreachable_storage_replica);
```

Comment:
We always throw "error_code_unreachable_storage_replica" when "requiredReplicas != ALL_REPLICAS" - this is because not all errors thrown above (captured by "replicaErrorCode") are re-triable (and "fetchKeys()" would cause the storage server to fail on such errors).

Reply:
Another thing to note is that this logic/throwing error is specifically geared towards the fetchKeys() use case. This might not work well if, say, the client/backup uses this setting. In general, my observation is that the error handling is done at various levels (by the load balancer itself, NativeAPI getxxxx() actors, invokers, like fetchKeys()), and there is not much uniformity on how these errors are handled (for example, the errors handled by Transaction::onError() are not exactly the same as the errors that fetchKeyCanRetry() tries to handle). So extending the code/various use cases to handle errors that we are introducing here is tricky, and it is hard to have confidence that we are getting it right in all cases.

Reply:
Agreed. We can follow up on what errors should be retried and where (i.e., defining a uniform way if possible)? For this PR, the current change looks good to me, since we know only fetchKeys() uses this setting.